src/macaron/slsa_analyzer/build_tool/poetry.py

```diff
@@ -115,4 +122,38 @@ def get_dep_analyzer(self, repo_path: str) -> DependencyAnalyzer:
         DependencyAnalyzer
             The DependencyAnalyzer object.
         """
+        # TODO: Implement this method.
         return NoneDependencyAnalyzer()
+
+    def get_build_dirs(self, repo_path: str) -> Iterable[Path]:
+        """Find directories in the repository that have their own build scripts.
+
+        This is especially important for applications that consist of multiple services.
+
+        Parameters
+        ----------
+        repo_path: str
+            The path to the target repo.
+
+        Yields
+        ------
+        Path
+            The relative paths from the repo path that contain build scripts.
+        """
+        config_paths: set[str] = set()
+        config_files = self.build_configs + self.package_lock
+        for build_cfg in config_files:
+            config_paths.update(glob.glob(os.path.join(repo_path, "**", build_cfg), recursive=True))
+
+        list_iter = iter(sorted(config_paths, key=lambda x: (str(Path(x).parent), len(Path(x).parts))))
+        try:
+            cfg_path = next(list_iter)
+            yield Path(cfg_path).parent.relative_to(repo_path)
+            while next_item := next(list_iter):
+                if str(Path(cfg_path).parent) in next_item:
```

Comment:
I would recommend we very careful of the `in` between two strings because the behavior might be implicit. For example, if `cfg_path` is `/home/boo` and `next_item` is `/home/foo/home/boo/` , `if str(Path(cfg_path).parent) in next_item:` will be True and we will ignore the valid `/home/foo/home/boo/`. 
I think we could use `.startswith` here ?

Reply:
That's a good point. I based this abstraction off the implementation in [BaseBuildTool.get_build_dirs()](https://github.com/oracle-samples/macaron/blob/add-python-as-a-supported-build-tool/src/macaron/slsa_analyzer/build_tool/base_build_tool.py#L162-L165), so this function would need updating too if that's the case.

Reply:
@behnazh-w How do you think about it?

Reply:
> That's a good point. I based this abstraction off the implementation in [BaseBuildTool.get_build_dirs()](https://github.com/oracle-samples/macaron/blob/add-python-as-a-supported-build-tool/src/macaron/slsa_analyzer/build_tool/base_build_tool.py#L162-L165), so this function would need updating too if that's the case.

That's not possible because first we sort the elements based on the [parent path](https://github.com/oracle-samples/macaron/blob/add-python-as-a-supported-build-tool/src/macaron/slsa_analyzer/build_tool/base_build_tool.py#L158). So `/home/foo/home/boo/` should never be filtered by `/home/boo`. But using `.startswith` would make it more explicit.

Regardless, I'm not sure if it makes sense to override this function to add `config_files = self.build_configs + self.package_lock`. As discussed, `poetry.lock` on its own cannot determine a build tool and `pyproject.toml` must always exist for poetry builds. So there shouldn't be any need to add `self.package_lock` here.

Reply:
@behnazh-w that's a good point. I had done it this way before we adjusted the config variable names, and now I agree that we don't need to override.

Reply:
Removed this override in [29cdd3a](https://github.com/oracle-samples/macaron/pull/67/commits/29cdd3a9e361a0f5f29c835805a2d9f186b92d0a).